capellambse/decl.py

```diff
@@ -0,0 +1,404 @@
+    """Load an instruction stream from a YAML file.
+
+    Parameters
+    ----------
+    file
+        An open file-like object, or a path or PathLike pointing to such
+        a file. Files are expected to use UTF-8 encoding.
+    """
+    if hasattr(file, "read"):
+        file = t.cast(t.IO[str], file)
+        ctx: t.ContextManager[t.IO[str]] = contextlib.nullcontext(file)
+    else:
+        assert not isinstance(file, t.IO)
+        ctx = open(file, encoding="utf-8")
+
+    with ctx as opened_file:
+        return yaml.load(opened_file, Loader=YDMLoader)
+
+
+def apply(model: capellambse.MelodyModel, file: FileOrPath) -> None:
+    """Apply a declarative modelling file to the given model.
+
+    Parameters
+    ----------
+    model
+        The model to apply the instructions to.
+    file
+        An open file-like object to read YAML instructions from, or a
+        path to such a file. Files will be read with UTF-8 encoding.
+
+    Notes
+    -----
+    This function is not transactional: If an exception occurs during
+    this function, the model will be left partially modified, with no
+    reliable way to know how much of the YAML input has been consumed.
+    It is therefore advised to reload or discard the model immediately
+    in these cases, to avoid working with an inconsistent state.
+
+    Even though the YAML layout suggests linear execution from top to
+    bottom, the actual order in which modifications are executed is
+    implementation defined. This is necessary to support promises with
+    ``!promise``, but reorderings are still possible even if no promises
+    are used in an input document.
+    """
+    instructions = collections.deque(load(file))
+    promises = dict[Promise, capellambse.ModelObject]()
+    deferred = collections.defaultdict[Promise, list[_FutureAction]](list)
+
+    while instructions:
+        instruction = instructions.popleft()
+        parent = instruction.pop("parent")
+        if isinstance(parent, UUIDReference):
+            parent = model.by_uuid(parent.uuid)
+        for op_type, apply_op in _OPERATIONS.items():
+            try:
+                op = instruction.pop(op_type)
+            except KeyError:
+                continue
+            for promise, outcome in apply_op(promises, parent, op):
+                if isinstance(outcome, dict):
+                    deferred[promise].append(outcome)
+                else:
+                    if promise in promises:
+                        raise ValueError(
+                            f"promise_id defined twice: {promise.identifier}"
+                        )
+                    promises[promise] = outcome
+                    instructions.extend(deferred.pop(promise, ()))
+        if instruction:
+            keys = ", ".join(instruction)
+            raise ValueError(f"Unrecognized keys in instruction: {keys}")
+    if deferred:
+        raise UnfulfilledPromisesError(frozenset(deferred))
+
+
+def _operate_create(
```

Comment:
IMO I'd like to have this public with a section about possible raised exceptions.

Why: Give the careful user the option to read about invalid/valid input through the declarative modeling language.